### ISSUE TITLE
Expect an error message always on fail event 

### ISSUE THREAD

**Issue by StylianosKoidakisSage** [badeball/cypress-cucumber-preprocessor#856]
### Current behavior
Using the following code i can not return cy as the preprocessor is expecting to return an error always

error message:
<img width="1043" alt="image" src="https://user-images.githubusercontent.com/84023810/195063251-330896f6-6484-4baf-badf-d0b5b84eceac.png">

```
  Cypress.on("fail", (error) => {
    if (error.message.includes(`input[name="email"]`)) {
      return cy;
    }
    throw error;
  });
```

### Desired behavior

I would like to be able on some failures to be able return cy and continue with the next steps. I know is not recommended but used to happen in the past and in some weird cases helps us to test our scenarios

### Versions

* **Cypress version**: "^10.9.0"
* **Preprocessor version**: "^13.0.2"
* **Node version**: v14.16.0

### Checklist

- [x] I've read the [FAQ](https://github.com/badeball/cypress-cucumber-preprocessor/blob/master/docs/faq.md).
- [x] I've read [Instructions for logging issues](https://github.com/badeball/cypress-cucumber-preprocessor/blob/master/CONTRIBUTING.md#instructions-for-logging-issues).
- [x] I'm not using `cypress-cucumber-preprocessor@4.3.1` (package name has changed and it is no longer the most recent version, see [#689](https://github.com/badeball/cypress-cucumber-preprocessor/issues/689)).

**Comment by badeball** [comment]
Can't reproduce, swallowing errors seems to work fine. You need to provide a full example, and as per above-mentioned instructions, don't post large bodies of code or screenshots. Make it easy for me by providing a cloneable repository that illustrates the issue.

**Comment by StylianosKoidakisSage** [comment]
@badeball  thank you very much for your answer, when I will find the time I will try to do a code sandbox that needs time to configure it. I have to mention that we use this part of code inside a command like so:

```
export const loginCommand = ({
  email,
  password,
}: {
  email: string;
  password: string;
}) => {
  Cypress.on("fail", (error) => {
    if (error.message.includes('input[name="email"]')) {
      return cy;
    }
    throw error;
  });
  cy.get('input[name="email"]', { timeout: 8000 })
    .type(email)
    .get('input[name="password"]')
    .type(password)
    .get('button[name="submit"]')
    .click()
    .get(".form")
    .should("not.exist");
};

```

My logic here is when I do not see the email input please continue with next step as there is not any error. Maybe the issue  exist because is a command and not a step.  When I return the `cy` the preprocessor expect an error

**Comment by StylianosKoidakisSage** [comment]
the issue happens when we use `cypress open` and we run different specs

**Comment by marlauda** [comment]
@badeball Hello, I face the same issue on a similar case.
The issue seems to occur as long as Cypress commands are called in [Cypress events](https://docs.cypress.io/api/events/catalog-of-events), and at least one attempt is failed.

It become problematic with [Cypress test retries mechanism](https://docs.cypress.io/guides/guides/test-retries), as it can lead in erroneously failed tests.

As requested, please find a reproductible example here : https://github.com/marlauda/demo_cucumber_preprocessor/tree/master

There are 2 tests available by tag :

- `intermittent_failure` => that test fail for every attempt except the last one, so always end up success as long as test retries is active.
- `no_failure` => that test never fail

If you run `intermittent_failure` alone, you will see in console logs that first attempt catch correctly the error. But the second attempt trigger the error described by @StylianosKoidakisSage, and so for the next attempts.
If you run `no_failure` alone, you will see that the test is success on every attempts
If you run both `intermittent_failure` and `no_failure`, you will see that the test retry mechanism is triggered on that last one, with the error described previously, for each execution.

Is it possible for you to have a look based on this new input?

Please feel free to answer if you need more information.

Thanks.

**Comment by badeball** [comment]
Hey, @marlauda. I've taken a look at your example and there's definitely something going on.

However, there's one thing I can't quite grasp with Cypress. If an error is swallowed using the code below, does or does not the rest of the test continue? Because it seems to me like the test will discontinue, the remaining steps be _not_ executed _and_ the test marked as passed by Cypress.

```js
Cypress.on("fail", () => {
  // Do nothing.
});
```

Does anyone know if this is intended behavior? The reason I ask is because this behavior can't really be modelled in a cucumber json report and thus can't supported.

**Comment by badeball** [comment]
> My logic here is when I do not see the email input please continue with next step as there is not any error. Maybe the issue exist because is a command and not a step. When I return the `cy` the preprocessor expect an error

Based on the comment [here](https://github.com/cypress-io/cypress/issues/5400#issuecomment-547059565), conditional testing wouldn't be possible using this approach, as not throwing in a fail-handler will still discontinue your test.

**Comment by badeball** [comment]
@marlauda, I've tried to play around with your example for a while now in order to isolate what's happened, but I couldn't. It seems to me like some of this stems from the fact thar you're invoking cypress commands from an event handler (`cy.task` in `test:after:run`), which is [unsupprted](https://github.com/cypress-io/cypress/issues/4823#issuecomment-586169559).

If I am to look more into this, I'm going to need a clearer example, which reproduces the issue _without_ resorting to something which itself is unsupported by cypress.

**Comment by marlauda** [comment]
Hello @badeball thanks a lot for your response.
Please find below answers to your questions :

> Hey, @marlauda. I've taken a look at your example and there's definitely something going on.
> 
> However, there's one thing I can't quite grasp with Cypress. If an error is swallowed using the code below, does or does not the rest of the test continue? Because it seems to me like the test will discontinue, the remaining steps be _not_ executed _and_ the test marked as passed by Cypress.
> 
> ```js
> Cypress.on("fail", () => {
>   // Do nothing.
> });
> ```
> 

> Does anyone know if this is intended behavior? The reason I ask is because this behavior can't really be modelled in a cucumber json report and thus can't supported.

You're right, catching that event, and never returning the error, will lead in always Success tests.
In my example, the error is sent, or a new one is triggered, so the test will never be a Success if an error occur.


> > My logic here is when I do not see the email input please continue with next step as there is not any error. Maybe the issue exist because is a command and not a step. When I return the `cy` the preprocessor expect an error
> 
> Based on the comment [here](https://github.com/cypress-io/cypress/issues/5400#issuecomment-547059565), conditional testing wouldn't be possible using this approach, as not throwing in a fail-handler will still discontinue your test.

Technically, conditional will work in there example, as it doesn't return an error in a specific case.
However, I think `return;` would be enough, not sure the `cy` object is necessary, and it would maybe avoid triggering the issue.

> @marlauda, I've tried to play around with your example for a while now in order to isolate what's happened, but I couldn't. It seems to me like some of this stems from the fact thar you're invoking cypress commands from an event handler (`cy.task` in `test:after:run`), which is [unsupprted](https://github.com/cypress-io/cypress/issues/4823#issuecomment-586169559).
> 
> If I am to look more into this, I'm going to need a clearer example, which reproduces the issue _without_ resorting to something which itself is unsupported by cypress.

The comment itself is from 2020, but if I look at the [official documentation](https://docs.cypress.io/api/events/catalog-of-events#Binding-to-Events), you're right that technically the `cy` object if supposed to be removed at this point.
But here all it calls is a task, which is meant to be executed anytime in the console, so I guess it's an exception to the behavior you pointed out.

My main concern is that it's working fine with `@badeball/cypress-cucumber-preprocessor` version < 15. 
I can't provide any other example about my case, as my investigation point out that this issue is precisely caused by the usage of cypress command in cypress event.

However, I will work on an example using @StylianosKoidakisSage initial comment, so maybe you will be able to help on there case (and by extension with some luck, mine).

**Comment by marlauda** [comment]
@badeball please find [here](https://github.com/marlauda/demo_cucumber_preprocessor/tree/demo_conditional_failure) an example based on the initial comment of @StylianosKoidakisSage

I reproduce it using `return cy;` or `return;`.

You just have to run `./node_modules/.bin/cypress run` to see the issue.

Hope this will help.

**Comment by badeball** [comment]
I've released v5.1.4, which I hope fixes this. Please check it out.

**Comment by marlauda** [comment]
It's working fine using your new version.

Thanks you.
